storage_controller/src/service.rs

```diff
@@ -5433,9 +5454,28 @@ impl Service {
             }
             AvailabilityTransition::ToActive => {
                 tracing::info!("Node {} transition to active", node_id);
+
+                let activate_node = {
+                    let locked = self.inner.read().unwrap();
+                    match locked.nodes.get(&node_id) {
+                        Some(n) => n.clone(),
+                        None => {
+                            return Err(ApiError::NotFound(
+                                anyhow::anyhow!("Node not registered").into(),
+                            ));
+                        }
+                    }
+                };
+
+                self.node_activate_reconcile(activate_node.clone(), nodes_lock)
```

Comment:
Is it safe to call this after having transitioned the node to active in `node_state_configure`? The doc comment explicitly calls this out as unsafe:

> This function must complete _before_ setting a `Node` to Active: once it is set to Active, other Reconcilers might communicate with the node, and these must not overlap with the work we do in this function.

Reply:
Nope, it's not. Updates to the observed state aren't fully sound at the moment (https://github.com/neondatabase/neon/issues/9124) and this would make it worse. Will fix.

Reply:
Done in https://github.com/neondatabase/neon/pull/9281/commits/02e1b064f22c89d821eb687f1edd281e5369fd59

Reply:
Thanks. Afaict it should be fine to reconcile on node activation before we've updated all node availabilities, since activation only cares about the state of the activated node and not any others that may have changed availability concurrently -- right?

Reply:
Right. `node_activate_reconcile` is not a great name (reconciliation is overloaded in this context).
It asks the nodes what tenant shards are present on it, updates in-mem state to reflect that, and, potentially, culls some tenant shards we don't recognize anymore.